README.md

```diff
@@ -1,289 +1,590 @@
+    ```
+1.  Mount a resource defintion value into the binding `Secret`
+    ```plain
+    “service.binding/uri”:
+      "path={.status.data.url}"
+    ```
+1.  Mount a resource definition value into the binding `Secret` with a different key
+    ```plain
+    “service.binding/uri":
+      "path={.status.data.connectionURL}”
+    ```
+1.  Mount the entries of a collection into the binding `Secret` selecting the key and value from each entry
+    ```plain
+    “service.binding/endpoints”:
+      "path={.status.bootstrap},elementType=sliceOfMaps,sourceKey=type,sourceValue=url"
+    ```
+
+## Mapping Existing Values to New Values
+
+Many applications will not be able to consume the secrets exposed by Provisioned Services directly.  Teams creating Provisioned Services do not know how their services will be consumed, teams creating Applications will not know what services will be provided to them, different language families have different idioms for naming and style, and more.  Users should have a way of describing a mapping from existing values to customize the provided entries to ones that are usable directly by their applications.  This specification is described as an extension to the [Service Binding](#service-binding) specification and assumes full compatibility with it.
+
+A Service Binding Resource **MAY** define a `.spec.mappings` which is an array of `Mapping` objects.  A `Mapping` object **MUST** define `name` and `value` entries.  The value of a `Mapping` **MAY** contain zero or more tokens beginning with `((`, ending with `))`, and encapsulating a binding `Secret` key name.  The value of this `Secret` entry **MUST** be substituted into the original `value` string, replacing the token.  Once all tokens have been substituted, the new `value` **MUST** be added to the `Secret` exposed to the resource represented by `application`.
+
+### Resource Type Schema
+
+```yaml
+apiVersion: service.binding/v1alpha1
+kind: ServiceBinding
+metadata:
+  name:         # string
+spec:
+  name:         # string, optional, default: .metadata.name
+
+  application:  # PodSpec-able resource ObjectReference
+    apiVersion: # string
+    kind:       # string
+    name:       # string
+    ...
+
+  service:      # Provisioned Service-able resource ObjectReference
+    apiVersion: # string
+    kind:       # string
+    name:       # string
+    ...
+
+  mapping:      # []Mapping, optional
+  - name:       # string
+    value:      # string
+  ...
+```
 
-### Mounting and injecting binding information
+### Example Resource
 
-#### Mounting data
-Implementations of this specification **MUST** mount the binding data into the consuming application container in the following location:
+```yaml
+apiVersion: service.binding/v1alpha1
+kind: ServiceBinding
+metadata:
+  name: online-banking-to-account-service
+spec:
+  name: account-service
+  kind: database
+  provider: vmware
 
+  application:
+    apiVersion: apps/v1
```

Comment:
+1 !

I'm been aligned to use `apiVersion` too!

Reply:
The fact that there's a core type called [`ObjectReference`](https://godoc.org/k8s.io/api/core/v1#ObjectReference) seems like a clear indication of the pattern to follow.

Reply:
umm.. was trying this out.

Unrelated to the spec:

`APIVersion` is marked 'optional' in `ObjectReference` - I wouldn't want it to be optional. Do you think I should still be able to use `ObjectReference` or is it OK to just define a custom struct? I remember now, this had pushed me towards Group, Version, Kind.
```
type ObjectFieldSelector struct {
	// Version of the schema the FieldPath is written in terms of, defaults to "v1".
	// +optional
	APIVersion string `json:"apiVersion,omitempty" protobuf:"bytes,1,opt,name=apiVersion"`
	// Path of the field to select in the specified API version.
	FieldPath string `json:"fieldPath" protobuf:"bytes,2,opt,name=fieldPath"`
}
```

Thoughts?

Reply:
Every field in ObjectReference is marked optional. Conceptually what we want is an object reference, I'm not concerned that is is the specific type, so along as that type is aligned with the fields on ObjectReference. For example, adding a label selector instead of name would require a different type, or at least embedding an ObjectReference inside the new struct.

Reply:
Resolved by c4a3221